### DISCUSSION TITLE
Workflow with pure python files?

### DISCUSSION THREAD

**Discussion by jorgesca** [benlubas/molten-nvim#163] [Q&A]
Hi, 
I am at a point now where I think I got all the bits and pieces working. I followed the guides, borrowed lots of code from @benlubas dotfiles. I can open a notebook, initialize a kernel, run some/all cells, etc. That's quite nice actually, although very complex to setup (at least for me). 
Now, I also tend to work with plain python files, mostly as a result of preferring neovim to a notebook in a browser (this predates molten and associated tools). The workflow then consisted in a split neovim window, the code in one pane and a terminal  running IPython on the other. I used iron.nvim then to communicate with the kernel. How would you do something like this with molten? Is it even a good idea to think about such a workflow? 
With the config I have now, I can open a python module, run a visual selection and see the output with portions of the code. But if attempt to run the whole file, I get a cryptic message saying ```-- No lines in buffer --```.  Maybe my setup still needs some tweaks here and there, but I thought about asking about your point of view regarding pure python  and molten before digging further.
BTW, thanks for your work. I know it's a collective effort with many parts needed, but you took the time and effort to create the glue needed so everything works.

**Comment by benlubas** [comment]
I use molten for regular python dev, but I don't think I've ever ran an entire file at once with it. I just run snippets of code. I use the floating windows in normal `.py` files b/c the virtual text just gets in the way. If you're unable to select the entire file contents and run it at once there might be a bug in molten.

I just tested, I only see this issue with virtual text:

With virtual text. I'm seeing that `virt_text_off_by_1 = true` will not let you run a cell that ends on the last line of the file. Or at least, it won't show you the output as virtual text (b/c there is no line to attach the output to. You can still run the entire file and then open the floating window.

If you want to run the entire file and see output as virtual text, you can just set `virt_text_off_by_1 = false` when you enter a python file, and then set it back to `true` when you leave. You can do that with an autocommand. There's an example in the notebook setup docs [here](https://github.com/benlubas/molten-nvim/blob/main/docs/Notebook-Setup.md#change-molten-settings-based-on-filetype)

**Comment by jorgesca** [comment]
Thanks for the hints. I think there might be something going on with the python file I was first using. It uses matplotlib's interactive feature for plots. I am not hundred percent certain about this, but when I tried with a simpler example that outputs only static images, I was able to see the output. I still have to play with the options as I see some quirks when using virtual text output, but at least when using the floating window I can see and enter the output. I'll close the question for now.

**Reply by benlubas** [reply]
Yeah interactive stuff generally means some type of javascript running. And I'm not too sure how to handle that b/c they require extra JS libraries. There's probably a way to get it to open in a web browser though. It's probably similar to #142.
